// @sap/llm-commons/build/cjs/langchain/embedding/amazon.cjs
'use strict';

var base = require('langchain/embeddings/base');
var client_amazon = require('../../client/amazon.cjs');
require('../../base-921f7eb6.cjs');
require('@sap-cloud-sdk/util');
require('../../index-2134ea38.cjs');
require('@sap/xsenv');
require('fs');
require('util');
require('axios');

var __defProp = Object.defineProperty;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __publicField = (obj, key, value) => {
  __defNormalProp(obj, typeof key !== "symbol" ? key + "" : key, value);
  return value;
};
class BTPAmazonTitanEmbedding extends base.Embeddings {
  constructor(fields) {
    var _a;
    super(__spreadValues({}, fields));
    __publicField(this, "btpAmazonTitanClient");
    __publicField(this, "deployment_id");
    this.deployment_id = (_a = fields == null ? void 0 : fields.deployment_id) != null ? _a : "amazon-titan-e1t-medium";
    this.btpAmazonTitanClient = new client_amazon.BTPAmazonTitanClient();
  }
  async embedQuery(document) {
    const res = await this.caller.callWithOptions(
      {},
      () => this.btpAmazonTitanClient.createEmbedding({
        deployment_id: this.deployment_id,
        inputText: document
      })
    );
    return res.embedding;
  }
  async embedDocuments(documents) {
    return Promise.all(documents.map((document) => this.embedQuery(document)));
  }
}

exports.BTPAmazonTitanEmbedding = BTPAmazonTitanEmbedding;
