// @sap/llm-commons/build/esm/langchain/embedding/amazon.mjs
import { Embeddings } from 'langchain/embeddings/base';
import { BTPAmazonTitanClient } from '../../client/amazon.mjs';
import '../../base-540b2ebe.mjs';
import '@sap-cloud-sdk/util';
import '../../index-924dd70f.mjs';
import '@sap/xsenv';
import 'fs';
import 'util';
import 'axios';

var __defProp = Object.defineProperty;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __publicField = (obj, key, value) => {
  __defNormalProp(obj, typeof key !== "symbol" ? key + "" : key, value);
  return value;
};
class BTPAmazonTitanEmbedding extends Embeddings {
  constructor(fields) {
    var _a;
    super(__spreadValues({}, fields));
    __publicField(this, "btpAmazonTitanClient");
    __publicField(this, "deployment_id");
    this.deployment_id = (_a = fields == null ? void 0 : fields.deployment_id) != null ? _a : "amazon-titan-e1t-medium";
    this.btpAmazonTitanClient = new BTPAmazonTitanClient();
  }
  async embedQuery(document) {
    const res = await this.caller.callWithOptions(
      {},
      () => this.btpAmazonTitanClient.createEmbedding({
        deployment_id: this.deployment_id,
        inputText: document
      })
    );
    return res.embedding;
  }
  async embedDocuments(documents) {
    return Promise.all(documents.map((document) => this.embedQuery(document)));
  }
}

export { BTPAmazonTitanEmbedding };
