// @sap/llm-commons/build/esm/langchain/embedding/btp.mjs
import { Embeddings } from 'langchain/embeddings/base';
import { BTPLLMProxyClient } from '../../client/btp.mjs';
import '../../index-924dd70f.mjs';
import '@sap-cloud-sdk/util';
import '@sap/xsenv';
import 'fs';
import 'util';
import '../../utils-8f77ab42.mjs';
import '../../client/ai21.mjs';
import '../../base-540b2ebe.mjs';
import 'axios';
import '../../client/alephalpha.mjs';
import '../../client/amazon.mjs';
import '../../client/anthropic.mjs';
import '../../client/cohere.mjs';
import '../../client/google.mjs';
import '../../client/huggingface.mjs';
import '../../client/openai.mjs';

var __defProp = Object.defineProperty;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __publicField = (obj, key, value) => {
  __defNormalProp(obj, typeof key !== "symbol" ? key + "" : key, value);
  return value;
};
class BTPLLMProxyEmbedding extends Embeddings {
  constructor(fields) {
    var _a, _b;
    super(__spreadValues({}, fields));
    __publicField(this, "BTPLLMProxyClient");
    __publicField(this, "deployment_id");
    __publicField(this, "batchSize", 512);
    __publicField(this, "stripNewLines", true);
    this.deployment_id = (_a = fields == null ? void 0 : fields.deployment_id) != null ? _a : "text-embedding-ada-002-v2";
    this.stripNewLines = (_b = fields == null ? void 0 : fields.stripNewLines) != null ? _b : this.stripNewLines;
    this.BTPLLMProxyClient = new BTPLLMProxyClient();
  }
  async embedDocuments(documents) {
    return Promise.all(documents.map((document) => this.embedQuery(document)));
  }
  async embedQuery(query) {
    const res = await this.caller.callWithOptions(
      {},
      () => this.BTPLLMProxyClient.createEmbedding({
        deployment_id: this.deployment_id,
        input: query
      })
    );
    return res.data[0].embedding;
  }
}

export { BTPLLMProxyEmbedding };
