// @sap/llm-commons/build/esm/langchain/embedding/google.mjs
import { GooglePaLMEmbeddings } from 'langchain/embeddings/googlepalm';
import { BTPGooglePaLMClient } from '../../client/google.mjs';
import '../../base-540b2ebe.mjs';
import '@sap-cloud-sdk/util';
import '../../index-924dd70f.mjs';
import '@sap/xsenv';
import 'fs';
import 'util';
import 'axios';

var __defProp = Object.defineProperty;
var __defProps = Object.defineProperties;
var __getOwnPropDescs = Object.getOwnPropertyDescriptors;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __spreadProps = (a, b) => __defProps(a, __getOwnPropDescs(b));
var __publicField = (obj, key, value) => {
  __defNormalProp(obj, typeof key !== "symbol" ? key + "" : key, value);
  return value;
};
class BTPGooglePaLMEmbedding extends GooglePaLMEmbeddings {
  constructor(fields) {
    var _a;
    super(__spreadProps(__spreadValues({}, fields), { apiKey: "dummy" }));
    __publicField(this, "btpGooglePaLMClient");
    __publicField(this, "deployment_id");
    this.deployment_id = (_a = fields == null ? void 0 : fields.deployment_id) != null ? _a : "gcp-textembedding-gecko-001";
    this.btpGooglePaLMClient = new BTPGooglePaLMClient();
  }
  async embedQuery(document) {
    return (await this.embedText([document]))[0];
  }
  async embedDocuments(documents) {
    return this.caller.callWithOptions({}, this.embedText.bind(this), documents);
  }
  async embedText(texts) {
    const cleanedText = texts.map((t) => t.replace(/\n/g, " "));
    const res = await this.btpGooglePaLMClient.embedText({
      deployment_id: this.deployment_id,
      instances: cleanedText.map((ct) => ({
        content: ct
      }))
    });
    return res.predictions.map((p) => p.embeddings.values);
  }
}

export { BTPGooglePaLMEmbedding };
